### PR TITLE
Serializable Endpoint/Bridge Annotations & Golden Gate Stage 

### PR DESCRIPTION
**Pull request by davidbiancolin**
To those reviewing, or looking at these changes, a high level summary:

1) Copied and modified FIRRTL JsonProtocol to support letting annotations register extra type hints
- Will try to upstream to FIRRTL

2) Stage implementation. in midas/stage
- Mostly copied from other existing Stage implementations. 
- Copied in GetIncludes from FIRRTL to use the modified JsonProtocol. Can be deleted after upstreaming JsonProtocol changes.
- The meat of the stage is implemented in the GoldenGateCompilerPhase which, against the spirit of the whole Stage/Phase enterprise, basically just wraps up what currently exists in Compiler.scala.  This phase also builds up the PLATFORM_CONFIG using the string-based convention we use in our generators
- Target and HostTransforms are passed to Golden Gate through a parameters object that is built up in this Phase

3) Modification of the Endpoint class hierarchies to support serialization 
- This requires the most scrutiny. See Endpoints.scala

4) Changes to existing Endpoints to support serializing their constructor arguments
  - FASED and PeekPoke required considerable reworking as they used constructs that could not be serialized

TODOs:

- [ ] Deprecate / Delete Compiler? 
- [ ] Clean up class hierarchies to remove in-memory versions

### CODE REVIEW

File: src/main/scala/midas/widgets/SerializationUtils.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  // Boxed types for different leaf chisel types we currently support
+  case class SerializableType(typeString: String)
+  val UIntType = SerializableType("UInt")
+  val SIntType = SerializableType("SInt")
+
+  case class SerializableField(name: String, tpe: SerializableType, fieldWidth: Int) {
+    def regenType(): Data = tpe match {
+      case UIntType => UInt(fieldWidth.W)
+      case SIntType => SInt(fieldWidth.W)
+      case _ => throw new Exception(s"Type string with no associated chisel type: ${tpe}")
```
[the remark this answers is not in the document]
Comment (davidbiancolin) [reply src/main/scala/midas/widgets/SerializationUtils.scala]: Yeah. But forcing a cast to Bool in the mean time isn't completely unreasonable right? I'll open an issue.

File: src/main/scala/midas/widgets/PeekPokeIO.scala [hunk 2]
```diff
@@ -183,10 +209,11 @@ class PeekPokeTargetIO(targetIO: Seq[(String, Data)], withReset: Boolean) extend
   override def cloneType = new PeekPokeTargetIO(targetIO, withReset).asInstanceOf[this.type]
 }
 
-class PeekPokeEndpoint(targetIO: Seq[(String, Data)], reset: Option[Bool]) extends BlackBox with IsEndpoint {
+class PeekPokeEndpoint(targetIO: Seq[(String, Data)], reset: Option[Bool]) extends BlackBox
+    with Endpoint[PeekPokeTokenizedIO, PeekPokeWidget] {
   val io = IO(new PeekPokeTargetIO(targetIO, reset != None))
-  val endpointIO = new PeekPokeEndpointIO(io)
-  def widget = (p: Parameters) => new PeekPokeIOWidget(endpointIO)(p)
+  val constructorArg = Some(PeekPokeKey(io))
```
[the remark this answers is not in the document]
Comment (davidbiancolin) [reply src/main/scala/midas/widgets/PeekPokeIO.scala]: You're talking about the constructorArg member name? or the use of Key?

File: src/main/scala/midas/widgets/SerializationUtils.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (albert-magyar) [reply src/main/scala/midas/widgets/SerializationUtils.scala]: Yeah that's fine

File: src/main/scala/midas/widgets/PeekPokeIO.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (albert-magyar) [reply src/main/scala/midas/widgets/PeekPokeIO.scala]: Oh I was saying that `constructorArg` is good. Very literal.

Using it in the annotations is probably not a bad idea.

**Pull request merged**